authentikos/authentikos.go

```diff
@@ -218,24 +247,45 @@ func loadClusterConfig() (*rest.Config, error) {
 
 // getOauthTokenCreator returns a function that creates/refreshes an oauth token.
 func getOauthTokenCreator(o options) (tokenCreator, error) {
+	var tokenCreator tokenCreator
+
 	clientOpts := []option.ClientOption{option.WithScopes(o.scopes...)}
 
 	if len(o.creds) > 0 {
 		clientOpts = append(clientOpts, option.WithCredentialsFile(o.creds))
 	}
 
+	// Reusing the client leverages the token source cache.
 	client, err := transport.Creds(context.Background(), clientOpts...)
-	if err != nil {
-		return nil, err
+
+	clientCreator := func(reuse bool) (*google.Credentials, error) {
+		if o.forceRefresh {
+			// Recreating the client invalidates the token source cache.
+			client, err = transport.Creds(context.Background(), clientOpts...)
+			printVerbose("force refreshing token\n", o.verbose)
+		}
+
+		return client, err
 	}
 
-	return func() ([]byte, error) {
+	tokenCreator = func(forceRefresh bool) ([]byte, error) {
+		client, err := clientCreator(forceRefresh)
+		if err != nil {
+			return nil, err
+		}
 		token, err := client.TokenSource.Token()
 		if err != nil {
 			return nil, err
 		}
+		nextRec := time.Now().Add(time.Duration(o.interval) * time.Minute)
+		if token.Expiry.Round(0).Add(-expiryDelta).Before(nextRec) {
+			// Force recreate the token if it will expire before the next reconciliation.
+			return tokenCreator(true)
```

Comment:
This would recurse forever if we repeatedly generate tokens that expire before the next reconciliation for some reason. Is that something that could happen?

Reply:
Added retry logic with a max retry limit.